### DISCUSSION TITLE
Conceptual packCells Questions

### DISCUSSION THREAD

**Discussion by gsangha1992** [UvaCsl/HemoCell#99] [General discussion]
Hi, 

My name is Gurneet and I'm a postdoc at University of Maryland. We recently installed Hemocell on our HPC to study red blood cell mechanics during exercise. I've been able to run the example simulations (e.g., pipeflow, curvedflow, preinlet) and I am now working on figuring out how to modify the hematocrit using packCells. For our first simulation study, we want to understand how hematocrit% alters forces on red blood cells in arterioles. I'll likely use pipeflow for this. I have a few conceptual/technical questions that have come up as I learn to use packCells . 

1.  How does hemocell scale the pipeflow geometry stl in the simulations? Is it scaling the geometry using the refDirN prescribed in the config.xml file? If so, does the geometry stl need to be generated at a certain resolution/voxels?

2. How do you determine the x,y,z dimensions of the packCells function? I know the dimensions are in microns. I'm confused because the function packs cells in a cube/rectangle, but the pipeflow simulations / arteries are using a cylinder geometry.  

3. I'm noticing that the curved flow example generates new cells at a specific pattern. In other words, it looks like the same 5-7 cells are being generated with the same initial positions. Can we use packCells to "release" red blood cells that all have different initial positions?

4. What is the upper threshold for Reynolds number? We would like to segment vessel geometries from 3D mouse vascular ultrasound data and calculate Reynolds number using pulsed wave velocity data. I noticed the documentation saying not to use a very high Reynolds number. Is there an upper limit to the simulations or is the upper limit determined by computation power?

Thank you for developing Hemocell! It's such a powerful tool that will complement our in vitro/animal/human studies.

**Comment by gzavo** [comment]
Hi @gsangha1992,

1. HemoCell does not take size information from the stl. Instead, you define the voxelization resolution (i.e., number of voxels along the reference axis). One voxel by default will correspond to 0.5 um. While you can also assign different size to a voxel, I would only recommend that if/when you are more familiar with the code and if you have some LBM and IBM experience. 
2. packCells always fills a rectangular domain and the cells generated outside of the actual simulated flow domain will not be used. This also means that if you generate a large rectangular domain packed with the necessary density, you can reuse it for multiple geometries that fit the rectangular bounding box.
3. No. packCells is a static initial condition. The curved flow uses a technique we refer to as "pre-inlet". You can read about the concept in 2D here: [https://doi.org/10.1016/j.compfluid.2018.04.025](https://doi.org/10.1016/j.compfluid.2018.04.025), and how it used in 3D here: [https://doi.org/10.1007/s10439-021-02870-4](https://doi.org/10.1007/s10439-021-02870-4).
4. The limit of numerical stability is typically bound by shear-rate, not Re. The reason is that the most unstable point is the force interaction between fluid and solid. We have run simulations with pulsatility at low Re (e.g. [https://doi.org/10.1080/10255842.2022.2034794](https://doi.org/10.1080/10255842.2022.2034794) ). You can model high shear rate (>10000 1/s) if it is constrained to a small section of the geometry, e.g. here we had 30000 1/s at the stenosis [https://doi.org/10.1098/rsif.2019.0148](https://doi.org/10.1098/rsif.2019.0148). If you have sustained shear, I think a few thousands is the max., and to my understanding that is already more than what other cellular codes are capable of.

**Comment by gsangha1992** [comment]
Hi @gzavo, 

Thanks for the feedback. This was helpful. I have a followup question for point 2. I modified the pipeflow config.xml to make refDirN 100 (original value was 50). I then used packCells to generate RBCs in a 50 50 200 rectangle. I used matlab to determine that the tube.stl length is 2x that of the diameter. However, the RBCs are not filling the whole volume in the length direction (see image below). I've tried using much greater length values with the same result. Am I missing something?

![Screen Shot 2023-05-15 at 9 30 31 PM](https://github.com/UvaCsl/HemoCell/assets/5167335/44194621-efa7-4832-ab58-c0eb8b776cab)

**Comment by gzavo** [comment]
Hi @gsangha1992,

The stl file is L=2*D, so if you pick refDirN=100, then the width is 200, and the domain you need to pack in [um] is 50 50 100.
(Of course it is possible to create an stl file that has a 1:4 diameter:length ratio).
